### PR TITLE
Add support for "dynamic" Immediate window locals

### PR DESCRIPTION
**Pull request by amcasey**
This is the third of a sequence of changes (see also 3dd1f2b and f2cac22).
In this change we start consuming the custom type info (effectively just
the flags of DynamicAttribute packed into a byte array) associated with
locals declared in the Immediate window.

TODO: The GetAliases API has not been implemented, but this functionality
will light up when it is.

### CODE REVIEW

**Comment by amcasey** [comment]
FYI @cston @ManishJayaswal @pnelsonmsft

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs [hunk 1]
```diff
@@ -109,7 +109,9 @@ private PlaceholderLocalSymbol LookupPlaceholder(string name)
                 return null;
             }
 
-            return CreatePlaceholderLocal(_typeNameDecoder, _containingMethod, new Alias(kind, id, id, typeName));
+            // The old API (GetObjectTypeNameById) doesn't return custom type info,
+            // but the new one (GetAliases) will.
+            return CreatePlaceholderLocal(_typeNameDecoder, _containingMethod, new Alias(kind, id, id, typeName, default(CustomTypeInfo)));
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs]: The comment is not necessary.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs [hunk 2]
```diff
@@ -123,6 +125,22 @@ private PlaceholderLocalSymbol LookupPlaceholder(string name)
             var type = typeNameDecoder.GetTypeSymbolForSerializedType(typeName);
             Debug.Assert((object)type != null);
 
+            var dynamicFlagsInfo = alias.CustomTypeInfo.ToDynamicFlagsCustomTypeInfo();
+            if (dynamicFlagsInfo.Any())
+            {
+                var flagsBuilder = ArrayBuilder<bool>.GetInstance();
+                dynamicFlagsInfo.CopyTo(flagsBuilder);
+                var dynamicType = DynamicTypeDecoder.TransformTypeWithoutCustomModifierFlags(
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs]: My suggestion to use BitArray originally was probably overkill. Would this be simpler if there were just two representations of the bits, as byte[] and bool[], with (static) methods to convert between them, much as you had previously?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Rewriters/LocalDeclarationRewriter.cs [hunk 3]
```diff
@@ -85,5 +92,51 @@ internal static BoundNode Rewrite(CSharpCompilation compilation, EENamedTypeSymb
...
+
+        private static BoundExpression GetCustomTypeInfoPayload(LocalSymbol local, CSharpSyntaxNode syntax, CSharpCompilation compilation, out bool hasCustomTypeInfoPayload)
+        {
+            var byteArrayType = new ArrayTypeSymbol(
+                compilation.Assembly,
+                compilation.GetSpecialType(SpecialType.System_Byte));
+
+            var flags = CSharpCompilation.DynamicTransformsEncoder.Encode(local.Type, customModifiersCount: 0, refKind: RefKind.None).ToArray();
+            var bytes = new DynamicFlagsCustomTypeInfo(new BitArray(flags)).GetCustomTypeInfoPayload();
+            hasCustomTypeInfoPayload = bytes != null;
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Rewriters/LocalDeclarationRewriter.cs]: Is bytes ever null? (It looks like DynamicTransformsEncoder.Encode always expects at least one true value.)

**Comment by cston** [comment]
LGTM

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs (hunk 1 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs]: I added it for PR purposes.  It will be deleted when the corresponding code is deleted (i.e. shortly).

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs (hunk 2 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Binders/PlaceholderLocalBinder.cs]: Let's discuss this and then make a separate PR.

**Comment by amcasey** [comment]
Oops, forgot to add Compilers for the DynamicTypeDecoder change.  @AlekseyTs @VSadov @gafter

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs [hunk 4]
```diff
@@ -63,7 +69,8 @@ private DynamicTypeDecoder(ImmutableArray<bool> dynamicTransformFlags, bool have
             {
                 return TransformTypeInternal(metadataType, containingModule.ContainingAssembly,
                     targetSymbolCustomModifierCount, targetSymbolRefKind, dynamicTransformFlags,
-                    haveCustomModifierFlags: true);
+                    haveCustomModifierFlags: true,
+                    checkLength: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs]: It is not obvious why "false" is the right value to pass here.

Comment (amcasey) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs]: Actually, I'm pretty sure that's wrong.  I'll see if I can write a test.

Comment (amcasey) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/DynamicTypeDecoder.cs]: I've added a test showing that it was wrong.

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Rewriters/LocalDeclarationRewriter.cs (hunk 3 above)
Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Rewriters/LocalDeclarationRewriter.cs]: I see the assert you're talking about, but I also know that code path is being hit, so I'll investigate.

Comment (amcasey) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Rewriters/LocalDeclarationRewriter.cs]: Aha, that assert only applies to the *other* overload of Encode.  The one I'm calling handles all-false (which seems to make way more sense than doing a pre-traversal to look for ```dynamic```).
